compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIROptimizer.java

```diff
@@ -312,86 +316,70 @@ private void reuseTempVariables(BIRFunction birFunction, OptimizerEnv funcOpEnv)
-            Set<BIRVariableDcl> secondLoadedSet = new HashSet<>();
-
             for (BIROperand birOperand : orderedTempVars) {
                 BIRVariableDcl variableDcl = birOperand.variableDcl;
-                List<BIRVariableDcl> reusableList = reusableVars.get(variableDcl);
-                // If given variable does not have any other variables to replace it we do not proceed.
-                if (reusableList == null || reusableList.isEmpty()) {
-                    secondLoadedSet.add(variableDcl);
-                    continue;
-                }
-                // Remove all removable vars with replaceable list.
-                reusableList.removeAll(removableVarVsReplacementVarMap.keySet());
                 if (!firstLoadedSet.contains(variableDcl)) {
                     firstLoadedSet.add(variableDcl);
-                    for (BIRVariableDcl birVariableDcl : reusableList) {
-                        if (!secondLoadedSet.contains(birVariableDcl)) {
-                            continue;
-                        }
-                        BIRVariableDcl currentUsingVar = reusableVsCurrentlyReplacedVarMap.get(birVariableDcl);
-                        // If no one use replaceable var then we can use it to replace given variable.
-                        if (currentUsingVar == null || secondLoadedSet.contains(currentUsingVar)) {
-                            // Collect selected removable var and it replacable var.
-                            removableVarVsReplacementVarMap.put(variableDcl, birVariableDcl);
-                            removableOpsMap.put(birOperand, birVariableDcl);
-                            reusableVsCurrentlyReplacedVarMap.put(birVariableDcl, variableDcl);
-                            break;
-                        }
+                    Set<BIRVariableDcl> reusableList = typeVsReusableVarsMap.computeIfAbsent(variableDcl.type,
+                            k -> new LinkedHashSet<>());
+                    if (reusableList.isEmpty()) {
+                        // No variables to reuse hence continue.
+                        continue;
                     }
+                    // We always get oldest reusable var.
+                    BIRVariableDcl reusableVar = reusableList.iterator().next();
+                    removableVarVsReplacementVarMap.put(variableDcl, reusableVar);
+                    removableOpsMap.put(birOperand, reusableVar);
+                    reusableList.remove(reusableVar);
+                    reusableList.remove(variableDcl);
+                    replaceableVarSet.add(reusableVar);
                     continue;
                 }
-                secondLoadedSet.add(variableDcl);
                 BIRVariableDcl reusableVar = removableVarVsReplacementVarMap.get(variableDcl);
-                // Same var can be used by multiple operands. we need to collect all and then replace the it
+                Set<BIRVariableDcl> reusableList = typeVsReusableVarsMap.get(variableDcl.type);
```

Comment:
Shall we have different name for the `reusableList` since we have another `reusableList` in the same function?

Reply:
It actually the same list.  This list is inside a type vs reusable list map. At first, we check if there is any reusable var with the same type to reuse. The second time when the variable is used, we can add that variable to the reusable list so any other variable can use.